internal/k8s/service/resolver.go

```diff
@@ -30,6 +30,7 @@ const (
 	ConsulServiceResolutionErrorType
 	GenericResolutionErrorType
 	NoResolutionErrorType
+	RefNotPermittedErrorType
```

Comment:
Thinking we may want to refactor this `ResolutionErrors` struct a bit.

As of now we basically store a bunch of errors that can happen during our resolution process here:

https://github.com/hashicorp/consul-api-gateway/blob/49e3d2fa58ebdef57122d4946c7b99f507400cfe/internal/k8s/service/resolver.go#L72-L79

When we go to set the final `ResolvedRefs` status in the above `OnBindFailed`, we flatten all of the errors that we've had and wrap the flattened error strings into a single "uber error":

https://github.com/hashicorp/consul-api-gateway/blob/49e3d2fa58ebdef57122d4946c7b99f507400cfe/internal/k8s/service/resolver.go#L121-L133

Right now, as written, the `Flatten()` call will result in this branch getting hit:

https://github.com/hashicorp/consul-api-gateway/blob/49e3d2fa58ebdef57122d4946c7b99f507400cfe/internal/k8s/service/resolver.go#L126

due to the lack of `RefNotPermittedErrorType` being taken into account in the error storage on the struct. Since we determine that a ref isn't permitted before resolving any other references on the route, I'm thinking it might make sense just to store the error directly on the struct and if it's not `nil` return the error with a `return RefNotPermittedErrorType, r.refError` in the call to `Flatten()`. Another potential way of handling this is by storing the error for the time being directly on the `K8sRoute` object rather than introducing a new error type here.

If anyone else has other suggestions too I'd be fine with that.

Reply:
I'm cool with implementing that, as long as there isn't going to be a potential situation where we might end up with more than one of them.

Reply:
Actually, that's a good point, since you can have multiple `backendRefs` with multiple required `ReferencePolicy` values, it does make sense to keep track of all of them and give feedback that a user could potentially be missing a few. So, I'm thinking that we can just extend `ResolutionErrors`, the storage arrays, `Add`, and the `Flatten` method. That's probably the most "correct" simplest way of doing things for now.

Thoughts @nathancoleman @mikemorris ?

Reply:
Agree